flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/logical/RewriteIntersectAll.scala

```diff
@@ -0,0 +1,124 @@
+  * {{{
+  *   SELECT c1
+  *   FROM (
+  *        SELECT replicate_row(min_count, c1)
+  *        FROM (
+  *             SELECT c1, If (vcol1_cnt > vcol2_cnt, vcol2_cnt, vcol1_cnt) AS min_count
+  *             FROM (
+  *                  SELECT   c1, count(vcol1) as vcol1_cnt, count(vcol2) as vcol2_cnt
+  *                  FROM (
+  *                       SELECT c1, true as vcol1, null as vcol2 FROM ut1
+  *                       UNION ALL
+  *                       SELECT c1, null as vcol1, true as vcol2 FROM ut2
+  *                       ) AS union_all
+  *                  GROUP BY c1
+  *                  HAVING vcol1_cnt >= 1 AND vcol2_cnt >= 1
+  *                  )
+  *             )
+  *         )
+  * }}}
+  */
+class RewriteIntersectAll extends ReplaceSetOpWithJoinRuleBase(
+  classOf[Intersect], "RewriteIntersectAll") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val intersect: Intersect = call.rel(0)
+    !intersect.isDistinct
+  }
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val intersect: Intersect = call.rel(0)
+    val left = intersect.getInput(0)
+    val right = intersect.getInput(1)
+
+    val fields = Util.range(intersect.getRowType.getFieldCount)
+
+    val builder1 = call.builder
+    val boolType = builder1.getTypeFactory.createSqlType(SqlTypeName.BOOLEAN)
+    val leftWithAddedVirtualCols = builder1
+        .push(left)
+        .project(
+          builder1.fields(fields) ++ Seq(
+            builder1.alias(builder1.literal(true), "vcol1"),
+            builder1.alias(builder1.getRexBuilder.makeNullLiteral(boolType), "vcol2")))
+        .build()
+
+    val builder2 = call.builder
+    val rightWithAddedVirtualCols = builder2
+        .push(right)
+        .project(
+          builder2.fields(fields) ++ Seq(
+            builder2.alias(builder2.getRexBuilder.makeNullLiteral(boolType), "vcol1"),
+            builder2.alias(builder2.literal(true), "vcol2")))
+        .build()
+
+    val builder = call.builder
+    builder
+        .push(leftWithAddedVirtualCols)
+        .push(rightWithAddedVirtualCols)
+        .union(true)
+        .aggregate(
+          builder.groupKey(builder.fields(fields)),
+          builder.count(false, "vcol1_count", builder.field("vcol1")),
+          builder.count(false, "vcol2_count", builder.field("vcol2")))
+        .filter(builder.and(
+          builder.call(
+            GREATER_THAN_OR_EQUAL,
+            builder.field("vcol1_count"),
+            builder.literal(1)),
+          builder.call(
+            GREATER_THAN_OR_EQUAL,
+            builder.field("vcol2_count"),
+            builder.literal(1))))
+        .project(Seq(builder.call(
+          IF,
+          builder.call(GREATER_THAN, builder.field("vcol1_count"), builder.field("vcol2_count")),
+          builder.field("vcol2_count"),
+          builder.field("vcol1_count"))) ++ builder.fields(fields))
+
+    call.transformTo(replicateRows(builder, intersect.getRowType, fields))
```

Comment:
please define a variable for `replicateRows(builder, intersect.getRowType, fields)` result, which could make `debug` more easier